## Filtrado con LOC and ILOC

Cuando queremos navegar por un dataFrame estas funciones permiten filtrar datos de manera más específica

In [10]:
import pandas as pd

best_sellers = pd.read_csv('../Data_Base/bestsellers_with_categories.csv', sep=',', header=0)
best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

Hagamos un slicing de los primeros 10 registros

In [11]:
best_sellers = best_sellers[:10]
best_sellers

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   
5      A Dance with Dragons (A Song of Ice and Fire)   
6  A Game of Thrones / A Clash of Kings / A Storm...   
7                     A Gentleman in Moscow: A Novel   
8      A Higher Loyalty: Truth, Lies, and Leadership   
9                          A Man Called Ove: A Novel   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction  
5       George R. R. Martin          4.4    12643     11  2011      Fiction  
6       George R. R. Martin          4.7    19735     30  2014      Fiction  
7               Amor Towles          4.7    19699     15  2017      Fiction  
8               James Comey          4.7     5983      3  2018  Non Fiction  
9           Fredrik Backman          4.6    23848      8  2016      Fiction

Filtremos por columnas, donde solo aparezcan Name and Author. Primero por Nombre

In [14]:
best_sellers['Name']

0                        10-Day Green Smoothie Cleanse
1                                    11/22/63: A Novel
2              12 Rules for Life: An Antidote to Chaos
3                               1984 (Signet Classics)
4    5,000 Awesome Facts (About Everything!) (Natio...
5        A Dance with Dragons (A Song of Ice and Fire)
6    A Game of Thrones / A Clash of Kings / A Storm...
7                       A Gentleman in Moscow: A Novel
8        A Higher Loyalty: Truth, Lies, and Leadership
9                            A Man Called Ove: A Novel
Name: Name, dtype: object

Ahora si por los dos

In [15]:
best_sellers[['Name', 'Author']]

Name                    Author
0                      10-Day Green Smoothie Cleanse                  JJ Smith
1                                  11/22/63: A Novel              Stephen King
2            12 Rules for Life: An Antidote to Chaos        Jordan B. Peterson
3                             1984 (Signet Classics)             George Orwell
4  5,000 Awesome Facts (About Everything!) (Natio...  National Geographic Kids
5      A Dance with Dragons (A Song of Ice and Fire)       George R. R. Martin
6  A Game of Thrones / A Clash of Kings / A Storm...       George R. R. Martin
7                     A Gentleman in Moscow: A Novel               Amor Towles
8      A Higher Loyalty: Truth, Lies, and Leadership               James Comey
9                          A Man Called Ove: A Novel           Fredrik Backman

### LOC 

Para que muestre todo

In [16]:
best_sellers.loc[:]

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   
5      A Dance with Dragons (A Song of Ice and Fire)   
6  A Game of Thrones / A Clash of Kings / A Storm...   
7                     A Gentleman in Moscow: A Novel   
8      A Higher Loyalty: Truth, Lies, and Leadership   
9                          A Man Called Ove: A Novel   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction  
5       George R. R. Martin          4.4    12643     11  2011      Fiction  
6       George R. R. Martin          4.7    19735     30  2014      Fiction  
7               Amor Towles          4.7    19699     15  2017      Fiction  
8               James Comey          4.7     5983      3  2018  Non Fiction  
9           Fredrik Backman          4.6    23848      8  2016      Fiction

Para filtrar del 0 al 4 incluyendolo

In [17]:
best_sellers.loc[0:4]

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

Vemos que tambien podemos usar *slicing* para que muestre lo mismo pero funciona un poco diferente a *loc*

In [18]:
best_sellers[0:5]

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

#### QUe mas se puede hacer con *loc*

In [24]:
best_sellers.loc[0:4, ['User Rating']] * (-1)

User Rating
0         -4.7
1         -4.6
2         -4.7
3         -4.7
4         -4.8

Ahora quiero comparar en todo el *data frame* cuando el autor es *Stephen King*

In [26]:
best_sellers.loc[:, ['Author']] == 'Stephen King'

Author
0   False
1    True
2   False
3   False
4   False
5   False
6   False
7   False
8   False
9   False

### iloc

Es muy parecido a loc. Lo que ocurre es que ahora la busqueda va a ser por indice, de hay el prefijo 'i'.

En el siguiente ejemplo le estoy diciendo que quiero todas las filas, pero las columnas del 0 a la 2  


In [30]:
best_sellers.iloc[:, 0:2]

Name                    Author
0                      10-Day Green Smoothie Cleanse                  JJ Smith
1                                  11/22/63: A Novel              Stephen King
2            12 Rules for Life: An Antidote to Chaos        Jordan B. Peterson
3                             1984 (Signet Classics)             George Orwell
4  5,000 Awesome Facts (About Everything!) (Natio...  National Geographic Kids
5      A Dance with Dragons (A Song of Ice and Fire)       George R. R. Martin
6  A Game of Thrones / A Clash of Kings / A Storm...       George R. R. Martin
7                     A Gentleman in Moscow: A Novel               Amor Towles
8      A Higher Loyalty: Truth, Lies, and Leadership               James Comey
9                          A Man Called Ove: A Novel           Fredrik Backman

En el siguiente ejemplo voy a comparar donde el autor sea *Amor Towels*

In [34]:
best_sellers.iloc[:,1] ==  'Amor Towles'

0    False
1    False
2    False
3    False
4    False
5    False
6    False
7     True
8    False
9    False
Name: Author, dtype: bool

O muy util para filtrar mucho mas los datos. Y una manera mucho mas sencilla para navegar entre todos estos datos

In [41]:
best_sellers.iloc[2:5, 1:4]

Author  User Rating  Reviews
2        Jordan B. Peterson          4.7    18979
3             George Orwell          4.7    21424
4  National Geographic Kids          4.8     7665

## Agregar o Eliminar datos con PANDAS

En primera instancia, buscaremos eleiminar algunas columnas del Data Frame, lo cual lo hacemos con el metodo *drop*

In [53]:
import pandas as pd

df_best_sellers = pd.read_csv('../Data_Base/bestsellers_with_categories.csv', sep= ',', header=0)
df_best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

A continuacion se borrara la columna *Genre*, junto con un parametro adicional: *axis*, recuerda 0 son filas, y 1 son columnas

### Una forma de eliminar datos

In [45]:
df_best_sellers.drop('Genre', axis=1).head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year  
0                  JJ Smith          4.7    17350      8  2016  
1              Stephen King          4.6     2052     22  2011  
2        Jordan B. Peterson          4.7    18979     15  2018  
3             George Orwell          4.7    21424      6  2017  
4  National Geographic Kids          4.8     7665     12  2019

Sin embargo, falta un parametro adicional para eliminar la columna de manera definitiva, porque como vemos, si volvemos a mostrar el Data Frame, veremos que persiste

In [49]:
df_best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

Y finalmente con el parametro *inplace* eliminamos la columna de manera definitiva

In [54]:
df_best_sellers.drop('Genre', axis=1, inplace=True)

In [55]:
df_best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year  
0                  JJ Smith          4.7    17350      8  2016  
1              Stephen King          4.6     2052     22  2011  
2        Jordan B. Peterson          4.7    18979     15  2018  
3             George Orwell          4.7    21424      6  2017  
4  National Geographic Kids          4.8     7665     12  2019

### Otra forma de eliminar datos

Y que para mi es la mas simple. Del Data Frame se procedera a eliminar la columna User Rating

In [56]:
df_best_sellers = df_best_sellers.drop('User Rating', axis=1)
df_best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  Reviews  Price  Year  
0                  JJ Smith    17350      8  2016  
1              Stephen King     2052     22  2011  
2        Jordan B. Peterson    18979     15  2018  
3             George Orwell    21424      6  2017  
4  National Geographic Kids     7665     12  2019

### Y ahora si, borremos filas

volvemos a cargar el data frame

In [67]:
import pandas as pd

df_best_sellers = pd.read_csv('../Data_Base/bestsellers_with_categories.csv', sep= ',', header=0)
df_best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

Borremos el primer registro con el metodo drop

In [68]:
df_best_sellers.drop(0, axis=0, inplace=True)

In [69]:
df_best_sellers.head()

Name  \
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   
5      A Dance with Dragons (A Song of Ice and Fire)   

                     Author  User Rating  Reviews  Price  Year        Genre  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction  
5       George R. R. Martin          4.4    12643     11  2011      Fiction

Y como ves la primera fila ya no es 0 sino 1

#### Ejemplo

Borrar los primeros 5 registros, partiendo que ya habiamos borrado el 1

In [74]:
df_best_sellers.drop(range(1,5), axis=0).head()

Name               Author  \
5      A Dance with Dragons (A Song of Ice and Fire)  George R. R. Martin   
6  A Game of Thrones / A Clash of Kings / A Storm...  George R. R. Martin   
7                     A Gentleman in Moscow: A Novel          Amor Towles   
8      A Higher Loyalty: Truth, Lies, and Leadership          James Comey   
9                          A Man Called Ove: A Novel      Fredrik Backman   

   User Rating  Reviews  Price  Year        Genre  
5          4.4    12643     11  2011      Fiction  
6          4.7    19735     30  2014      Fiction  
7          4.7    19699     15  2017      Fiction  
8          4.7     5983      3  2018  Non Fiction  
9          4.6    23848      8  2016      Fiction

### ¿Y como agregar nuevas filas o columnas?

Volvemos a leer el data frame, y añadir una nueva columna con NAN


**NaN** stands for Not A Number and is one of the common ways to represent the missing value in the data

In [76]:
import pandas as pd
import numpy as np

df_best_sellers = pd.read_csv('../Data_Base/bestsellers_with_categories.csv', sep= ',', header=0)
df_best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

In [79]:
df_best_sellers['New Column'] = np.NaN
df_best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  \
0                  JJ Smith          4.7    17350      8  2016  Non Fiction   
1              Stephen King          4.6     2052     22  2011      Fiction   
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction   
3             George Orwell          4.7    21424      6  2017      Fiction   
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction   

   New Column  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

Verifiquemos con el metodo .shape la cantidad de datos

In [80]:
df_best_sellers.shape

(550, 8)

#### Ejemplo

Agregar una nueva columna llamada *Rango* a partir de un arreglo de numpy

In [83]:
data = np.arange(0,df_best_sellers.shape[0])

df_best_sellers['Rango'] = data

In [84]:
df_best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  \
0                  JJ Smith          4.7    17350      8  2016  Non Fiction   
1              Stephen King          4.6     2052     22  2011      Fiction   
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction   
3             George Orwell          4.7    21424      6  2017      Fiction   
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction   

   New Column  Rango  
0         NaN      0  
1         NaN      1  
2         NaN      2  
3         NaN      3  
4         NaN      4

#### Finalmente agregamos filas o ROWS

Hay muchas opciones pero la mas facil y mas eficiente en memoria y demas, es el metodo *append*:

Por ejemplo si usamos el metodo sobre el mismo: tenemos ahora el doble de registros como era obvio

In [85]:
df_best_sellers.append(df_best_sellers)

/tmp/ipykernel_379/1746177009.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_best_sellers.append(df_best_sellers)


Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  \
0                    JJ Smith          4.7    17350      8  2016  Non Fiction   
1                Stephen King          4.6     2052     22  2011      Fiction   
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction   
3               George Orwell          4.7    21424      6  2017      Fiction   
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction   
..                        ...          ...      ...    ...   ...          ...   
545               Jeff Kinney          4.9     9413      8  2019      Fiction   
546               Jen Sincero          4.7    14331      8  2016  Non Fiction   
547               Jen Sincero          4.7    14331      8  2017  Non Fiction   
548               Jen Sincero          4.7    14331      8  2018  Non Fiction   
549               Jen Sincero          4.7    14331      8  2019  Non Fiction   

     New Column  Rango  
0           NaN      0  
1           NaN      1  
2           NaN      2  
3           NaN      3  
4           NaN      4  
..          ...    ...  
545         NaN    545  
546         NaN    546  
547         NaN    547  
548         NaN    548  
549         NaN    549  

[1100 rows x 9 columns]

Como se observa el metodo *append* pronto sera depreciado y sera sustituido por el metodo *concat* el cual todavia no domino, como se observa a continuacion

In [100]:
import pandas as pd
import numpy as np

df_best_sellers = pd.read_csv('../Data_Base/bestsellers_with_categories.csv', sep= ',', header=0)
df_best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

In [131]:
my_dict = {'Name':'Cien años de Soledad', 
'Author':'Gabriel García Márquez', 
'User Rating': [4.8], 
'Reviews':[20000],
'Year':[1967], 
'Price':[25]}


In [133]:
my_dict = pd.DataFrame(my_dict)
my_dict

Name                  Author  User Rating  Reviews  Year  \
0  Cien años de Soledad  Gabriel García Márquez          4.8    20000  1967   

   Price  
0     25

In [134]:
pd.concat((df_best_sellers, my_dict))

Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   
0                                 Cien años de Soledad   

                       Author  User Rating  Reviews  Price  Year        Genre  \
0                    JJ Smith          4.7    17350      8  2016  Non Fiction   
1                Stephen King          4.6     2052     22  2011      Fiction   
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction   
3               George Orwell          4.7    21424      6  2017      Fiction   
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction   
..                        ...          ...      ...    ...   ...          ...   
546               Jen Sincero          4.7    14331      8  2016  Non Fiction   
547               Jen Sincero          4.7    14331      8  2017  Non Fiction   
548               Jen Sincero          4.7    14331      8  2018  Non Fiction   
549               Jen Sincero          4.7    14331      8  2019  Non Fiction   
0      Gabriel García Márquez          4.8    20000     25  1967          NaN   

    New Column  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  
..         ...  
546        NaN  
547        NaN  
548        NaN  
549        NaN  
0          NaN  

[551 rows x 8 columns]

#### Datos nulos y Pandas

Pandas implemento valores nulos dentro del mismo pandas para no tener que importar NumPy, podemos usar pd.NA y es lo mismo que np.nan

In [104]:
import pandas as pd

df_best_sellers = pd.read_csv('../Data_Base/bestsellers_with_categories.csv', sep= ',', header=0)
df_best_sellers.head()

df_best_sellers['New Column'] = pd.NA
df_best_sellers.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  \
0                  JJ Smith          4.7    17350      8  2016  Non Fiction   
1              Stephen King          4.6     2052     22  2011      Fiction   
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction   
3             George Orwell          4.7    21424      6  2017      Fiction   
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction   

  New Column  
0       <NA>  
1       <NA>  
2       <NA>  
3       <NA>  
4       <NA>

## Manjeo de Datos Nulos

Los datos nulos son un dolor de cabeza en el mundo de Data Science, y aqui vamos a ver como tratrar con ellos.

Como punto de partida, se crea un dicccionario, y a partir de el se crea un objeto de tipo DataFrame

In [135]:
import pandas as pd
import numpy as np

my_dict = {'Col1':[1,2,3,np.nan],
'Col2': [4, np.nan, 6,7],
'Col3': ['a','b','c', None]}

df = pd.DataFrame(my_dict) 
df.head()

Col1  Col2  Col3
0   1.0   4.0     a
1   2.0   NaN     b
2   3.0   6.0     c
3   NaN   7.0  None

El dataset que acabamos de crear es muy comun verlo, con valores incompletos y nulos. El primer metodo seria *is_null*, que retorna True si es un valor None or NaN

In [136]:
df.isnull()

Col1   Col2   Col3
0  False  False  False
1  False   True  False
2  False  False  False
3   True  False   True

La mejor forma de hacerlo es multiplicaarlo por 1, y ya veras porque

In [138]:
df.isnull()*1

Col1  Col2  Col3
0     0     0     0
1     0     1     0
2     0     0     0
3     1     0     1

Una vez los hallamos identifado, y para decirlo de alguna forma, normalizar los datos, colocarles una etiqueta de 'Missing' con el metodo *fillna*

In [139]:
df.fillna('Missing')

Col1     Col2     Col3
0      1.0      4.0        a
1      2.0  Missing        b
2      3.0      6.0        c
3  Missing      7.0  Missing

Pero eso no es todo, puedo usar *fillna* y llenarlos con valores numericos, por ejemplo la media, la moda, la STV, etc

Para este caso me indicara un warning porque la columna 3 obviamente no son valores numericos

In [141]:
df.fillna(df.mean())

/tmp/ipykernel_379/634187881.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df.fillna(df.mean())


Col1      Col2  Col3
0   1.0  4.000000     a
1   2.0  5.666667     b
2   3.0  6.000000     c
3   2.0  7.000000  None

Otra forma de tratar con valores NaN es interpolarlos, es decir, se le indicara que es una serie de datos, y tratara de buscar el numero que hace falta para completar la serie: *interpolate*

In [142]:
df.interpolate() 

Col1  Col2  Col3
0   1.0   4.0     a
1   2.0   5.0     b
2   3.0   6.0     c
3   3.0   7.0  None

Y una ultima forma de tratar con los NaN, y muy comun entre los cientificos de datos, es eliminarlos: *dropna*. Esto como siempre es usando el mejor criterio. Si por ejemplo son tan pocos los valores NaN con relacion al tamaño del DataFrame, pues mejor borrarlos 

In [143]:
df.dropna()

Col1  Col2 Col3
0   1.0   4.0    a
2   3.0   6.0    c

## Filtrado por Condiciones

Al igual que en Numpy, tambien podemos filtrar en Pandas

In [149]:
import pandas as pd

df = pd.read_csv('../Data_Base/bestsellers_with_categories.csv')
print(df.shape)
df.head()

(550, 7)


Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

Por ejemplo donde al año sea mayor o igual a 2015

In [151]:
df['Year'] >= 2017

0      False
1      False
2       True
3       True
4       True
       ...  
545     True
546    False
547     True
548     True
549     True
Name: Year, Length: 550, dtype: bool

Como ves es muy parecido a como lo veniamos haciendo con Numpy, porque Pandas esta basado en Numpy.

Usemos la condicion anterior para filtrar los datos

In [152]:
mayor_igual_2017 = df['Year'] >= 2017

df[mayor_igual_2017]

Name  \
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
7                       A Gentleman in Moscow: A Novel   
8        A Higher Loyalty: Truth, Lies, and Leadership   
..                                                 ...   
544                                             Wonder   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
7                 Amor Towles          4.7    19699     15  2017      Fiction  
8                 James Comey          4.7     5983      3  2018  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
544             R. J. Palacio          4.8    21625      9  2017      Fiction  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[150 rows x 7 columns]

Y como ves ahora tenemos 150 registros a comparacion de los 550 originales. Tambien se puede implementar directamente...

In [153]:
df[df['Year'] >= 2017]

Name  \
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
7                       A Gentleman in Moscow: A Novel   
8        A Higher Loyalty: Truth, Lies, and Leadership   
..                                                 ...   
544                                             Wonder   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
7                 Amor Towles          4.7    19699     15  2017      Fiction  
8                 James Comey          4.7     5983      3  2018  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
544             R. J. Palacio          4.8    21625      9  2017      Fiction  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[150 rows x 7 columns]

#### Ejemplo

Filtrar los datos de tal forma que solo aparezcan los que no sean de ciencia ficcion

In [154]:
df[df['Genre'] != 'Fiction']

Name  \
0                        10-Day Green Smoothie Cleanse   
2              12 Rules for Life: An Antidote to Chaos   
4    5,000 Awesome Facts (About Everything!) (Natio...   
8        A Higher Loyalty: Truth, Lies, and Leadership   
11   A Patriot's History of the United States: From...   
..                                                 ...   
539  Women Food and God: An Unexpected Path to Almo...   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
8                 James Comey          4.7     5983      3  2018  Non Fiction  
11           Larry Schweikart          4.6      460      2  2010  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
539               Geneen Roth          4.2     1302     11  2010  Non Fiction  
546               Jen Sincero          4.7    14331      8  2016  Non Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[310 rows x 7 columns]

Y tambien, sumando complejidad integrando mas condiciones.

#### Ejemplo 

Encontrar todos los registros donde el genero se Fiction, y el año de publicacion mayor a 2016. Encontrando 65 filas que cumplen con la condicion

In [156]:
fiction = df['Genre'] == 'Fiction'

mayor_2016 = df['Year'] > 2016


df[fiction & mayor_2016]

Name             Author  \
3                               1984 (Signet Classics)      George Orwell   
7                       A Gentleman in Moscow: A Novel        Amor Towles   
10                           A Man Called Ove: A Novel    Fredrik Backman   
13                    A Wrinkle in Time (Time Quintet)  Madeleine L'Engle   
40            Brown Bear, Brown Bear, What Do You See?    Bill Martin Jr.   
..                                                 ...                ...   
509                              To Kill a Mockingbird         Harper Lee   
529  What Should Danny Do? (The Power to Choose Ser...          Adir Levy   
534                            Where the Crawdads Sing        Delia Owens   
544                                             Wonder      R. J. Palacio   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)        Jeff Kinney   

     User Rating  Reviews  Price  Year    Genre  
3            4.7    21424      6  2017  Fiction  
7            4.7    19699     15  2017  Fiction  
10           4.6    23848      8  2017  Fiction  
13           4.5     5153      5  2018  Fiction  
40           4.9    14344      5  2017  Fiction  
..           ...      ...    ...   ...      ...  
509          4.8    26234      7  2019  Fiction  
529          4.8     8170     13  2019  Fiction  
534          4.8    87841     15  2019  Fiction  
544          4.8    21625      9  2017  Fiction  
545          4.9     9413      8  2019  Fiction  

[65 rows x 7 columns]

Si queremos negar la condicion, y para reutilizar codigo, usamos el simbolo **~** para negarla. POr ejemplo encontremos los libros de Ficcion que no sean mayores a 2016.  

In [157]:
df[fiction & ~mayor_2016]

Name               Author  \
1                                    11/22/63: A Novel         Stephen King   
5        A Dance with Dragons (A Song of Ice and Fire)  George R. R. Martin   
6    A Game of Thrones / A Clash of Kings / A Storm...  George R. R. Martin   
9                            A Man Called Ove: A Novel      Fredrik Backman   
20                         All the Light We Cannot See        Anthony Doerr   
..                                                 ...                  ...   
538  Winter of the World: Book Two of the Century T...          Ken Follett   
540                                             Wonder        R. J. Palacio   
541                                             Wonder        R. J. Palacio   
542                                             Wonder        R. J. Palacio   
543                                             Wonder        R. J. Palacio   

     User Rating  Reviews  Price  Year    Genre  
1            4.6     2052     22  2011  Fiction  
5            4.4    12643     11  2011  Fiction  
6            4.7    19735     30  2014  Fiction  
9            4.6    23848      8  2016  Fiction  
20           4.6    36348     14  2014  Fiction  
..           ...      ...    ...   ...      ...  
538          4.5    10760     15  2012  Fiction  
540          4.8    21625      9  2013  Fiction  
541          4.8    21625      9  2014  Fiction  
542          4.8    21625      9  2015  Fiction  
543          4.8    21625      9  2016  Fiction  

[175 rows x 7 columns]